File: 2021/02/part1/cpp/main.cc
#include <iomanip>
#include <iostream>

#include <algorithm>
#include <memory>
#include <numeric>

#include <bitset>
#include <map>
#include <optional>
#include <string>
#include <variant>
#include <vector>

#include <cassert>
#include <cctype>

enum class Direction
{
  UP      = -1,
  FORWARD = 0,
  DOWN    = +1,
};

std::string
to_string(Direction d)
{
  switch (d)
  {
    case Direction::DOWN: return "down";
    case Direction::FORWARD: return "forward";
    case Direction::UP: return "up";
    default:
    {
      std::cerr << "unknown direction value '" << int(d) << "'" << std::endl;
      exit(-1);
    }
  }
}

Direction
from_string(const std::string& s)
{
  if (s == "down")
  {
    return Direction::DOWN;
  }
  else if (s == "up")
  {
    return Direction::UP;
  }
  else if (s == "forward")
  {
    return Direction::FORWARD;
  }
  else
  {
    throw std::invalid_argument("not a valid direction input");
  }
}

using Magnitude = int;

struct Movement
{
  Direction direction;
  Magnitude magnitude;
};

std::istream&
operator>>(std::istream& is, Movement& m)
{
  std::string direction_str;
  is >> direction_str;
  try
  {
    m.direction = from_string(direction_str);
    is >> m.magnitude;
  }
  catch (const std::invalid_argument& exception)
  {
    is.setstate(std::ios::failbit);
  }

  return is;
}

std::ostream&
operator<<(std::ostream& os, const Movement& m)
{
  os << to_string(m.direction) << ' ' << m.magnitude;

  return os;
}

using Depth = int;
using Horiz = int;
struct Position
{
  Horiz horiz = 0;
  Depth depth = 0;
  void  apply(const Movement& movement)
  {
    // (relying on up=(-1), down=(+1))
    // change depth on up or down
    auto delta_depth = int(movement.direction) * int(movement.magnitude);
    this->depth += delta_depth;

    // if no depth change then either movement is forward or magnitude is 0;
    // in either case we can simply apply the magnitude forward.
    this->horiz += (delta_depth == 0) ? movement.magnitude : 0;
  }
};
std::ostream&
operator<<(std::ostream& os, const Position& p)
{
  os << '(' << "horiz:" << p.horiz << ',' << "depth:" << p.depth << ')';
  return os;
}

class Submarine
{
public:
  void     move(const Movement& movement) { this->pos.apply(movement); }
  Position get_pos(void) const { return this->pos; };

private:
  Position pos;
};

int
main(void)
{
  // get values
  std::vector<Movement> movements;
  {
    Movement movement;
    while (std::cin >> movement) { movements.push_back(movement); }
  }

  Submarine sub;
  for (const auto& movement : movements) { sub.move(movement); }
  const auto& pos           = sub.get_pos();
  const auto [horiz, depth] = pos;

  std::cout << "final position: " << pos << std::endl;
  std::cout << "multiplied: " << (horiz * depth) << std::endl;
}

File: 2021/02/part1/rust/src/main.rs
use std::io;
use std::str::FromStr;
use std::vec::Vec;
use strum_macros::Display;
use strum_macros::EnumString;

struct Sub {
    pos: Position,
}

impl Sub {
    fn apply(&mut self, movement: &Movement) {
        match movement.direction {
            Direction::Up => self.pos.depth -= movement.magnitude,
            Direction::Down => self.pos.depth += movement.magnitude,
            Direction::Forward => self.pos.horiz += movement.magnitude,
        }
    }
}

struct Position {
    horiz: i32,
    depth: i32,
}

struct Movement {
    direction: Direction,
    magnitude: i32,
}

#[derive(Display, EnumString)]
#[strum(serialize_all = "snake_case")]
enum Direction {
    Up,
    Down,
    Forward,
}

fn main() {
    // parse movements into vector
    let mut movements: Vec<Movement> = Vec::new();
    loop {
        let mut input = String::new();

        let bytes_read = io::stdin()
            .read_line(&mut input)
            .expect("Failed to read line");
        let is_eof = bytes_read == 0;
        if is_eof {
            break;
        }

        let mut tokens_iter = input.trim().split_whitespace();
        let direction = tokens_iter.next().unwrap();
        let magnitude = tokens_iter.next().unwrap();

        let direction = Direction::from_str(direction).unwrap();
        let magnitude: i32 = match magnitude.parse() {
            Ok(num) => num,
            Err(_) => continue,
        };

        let movement: Movement = Movement {
            direction,
            magnitude,
        };

        movements.push(movement)
    }

    // apply movements to sub
    let mut sub: Sub = Sub {
        pos: Position { horiz: 0, depth: 0 },
    };
    for movement in movements.iter() {
        sub.apply(movement);
    }

    // emit result
    println!("final pos: (horiz:{},depth:{})", sub.pos.horiz, sub.pos.depth);
    println!("multiplied: {}", (sub.pos.horiz * sub.pos.depth));
}
